Create Table View [PTC-160]

voidus opened the pull request:
https://mayflower-osl.atlassian.net/browse/PTC-160

Commit:
Added projectUserMinutesAction

This retrieves the minutes booked in a specific time separated by
project and users

Commit:
Also send the project title and user name in projectUserMinutesAction

Commit:
Added getProjectUserMinutes to the timecard client model

Commit:
Move the buttons below the header, just above the table/graph

This emphasizes that they belong to the graph only.

Commit:
Added projectUserTimeTable to the statistics page

exi commented on phprojekt/application/Timecard/Controllers/IndexController.php:
```diff
@@ -432,6 +432,40 @@ public function workBalanceByDayAction()
 
+    public function projectUserMinutesAction()
+    {
+        $startDate  = new DateTime($this->getRequest()->getParam('start'));
```
input validation anyone?
i don't know whether fail by exception is the way to go

voidus replied:
Fixed in 7b0c55

exi commented:
I don't think that the arrangement of the table is ok like that.
i would lay it out like that:
|----------------------------------|
| Project      | Time        |
|-----------------|----------------|
| Project 1   | 3h            |
| Project 2   | 4h            |
|----------------------------------|

displaying the username ist entirely useless as the ticket itself only describes the user perspective and not the team perspective. 
"As a user, I want to see _my_ total time booked per project for this month in a table. "

As far as i understand the task, we display the team view in another table

voidus commented:
This layout was discussed by mail some time ago. I think we show the user name to make it consistent with the team manager view that will follow.

Commit:
Added IndexController->getDateParam(...) to validate isodates

This will throw an Phprojekt_Exception_InvalidParameter if the parameter is
malformatted.

Commit:
use _getDateParam to validate isodate parameters

Commit:
simplify regexp a little

Commit:
correctly check the return value of preg_match

Commit:
Move the hairy sql code to the timecard model

Commit:
remove extra whitespace

exi commented on phprojekt/application/Timecard/Models/Timecard.php:
```diff
@@ -572,4 +572,27 @@ private function _checkSaveRights()
         }
     }
+
+    public static function getMinutesByUserAndProject($userIds, $startDate, $endDate)
```
"getProjectMinutesByUsers"?

voidus replied:
done

exi commented on phprojekt/application/Calendar2/Controllers/IndexController.php:
```diff
@@ -107,18 +107,10 @@ public function jsonListAction()
+        $dateEnd   = $this->_getDateParam('dateEnd');
         $userId    = $this->getRequest()->getParam('userId', (int) Phprojekt_Auth_Proxy::getEffectiveUserId());
 
-        if (!Cleaner::validate('isoDate', $dateStart)) {
```
I think it is a bad idea to not use the cleaner.
This will lead to duplicate code.

voidus replied:
I changed the _getDateParam method to use the Cleaner. This avoids the explicit check in all controller methods.

exi commented on phprojekt/htdocs/phpr/Statistics/ProjectUserTimeTable.js:
```diff
@@ -0,0 +1,89 @@
+                    };
+                }
+
+                byUserId[entry.user_id][entry.project_id] = timehelper.minutesToHMString(entry.minutes);
```
look at line 44 and assert whether there actually is a project_id field in byUserId[entry.user_id] ;)

voidus replied:
The entry objects are created in Timecard_Models_Timecard::getProjectMinutesByUser. It contains project_id, user_id, minutes, project and user fields.

exi replied:
that might be right, but in this block:

<pre>
                if (!byUserId.hasOwnProperty(entry.user_id)) {
                    byUserId[entry.user_id] = {
                        user: entry.user,
                        id: entry.user_id
                    };
                }
</pre>

you create byUserId[entry.user_id] with attributes user and id.
later you access byUserId[entry.user_id][entry.project_id] which accesses the entry.project_id attribute of byUserId[entry.user_id] which does not exist

voidus replied:
Now I understand. byUserId[entry.user_id] doesn't habe an entry.project_id key, but that's not an issue as I am assigning it here rather than accessing it.

exi commented on phprojekt/htdocs/phpr/Statistics/ProjectUserTimeTable.js:
```diff
@@ -0,0 +1,89 @@
+            });
+            this.own(grid);
+            grid.placeAt(this.domNode);
+            grid.startup();
```
placeAt automatically calls startup

voidus replied:
fixed

Commit:
Remove the call to startup after placeAt

It's called automatically

Commit:
Rename getMinutesByUserAndProject to getProjectMinutesByUsers

This stresses the point that the users are given as parameters, while the
function returns values for all projects.

There was also a typo in the method call that was corrected by this

Commit:
Use Cleaner to validate date parameters

Commit:
Merge remote-tracking branch 'mayflower/next' into projectTable

* mayflower/next: (37 commits)
  add missing index.json.phtml
  remove getGlobalModules call in frontend
  fix startup sequence of the javascript
  Fix an error in Project/Sql/Db.json
  use the css class "button" instead of "submitButton" for our buttons
  use single quotation marks consistently
  Fix SQL string creation for create table statements
  Fix typo in SQL
  add destroyed checks to summary block
  Update composer.lock
  remove debugger;
  Remove unused code
  PHPROJEKT-157: Database for project managers
  don't force dayblocks closed if they are not in the last 2 days
  show unfinished bookings in red
  cleanup dayblock code
  remove dead code from dayblock
  add missing open attribute to dayblock
  rename bookinglist variables
  set project table to use autoheight
  ...

Conflicts:
	phprojekt/htdocs/phpr/StatisticsView.js
	phprojekt/htdocs/phpr/template/statisticsView.html
